doc/rst/technotes/initequals.rst

```diff
@@ -0,0 +1,268 @@
+through the expression ``this.type``. The compiler-generated ``init=`` for
+type ``R`` looks like:
+
+.. code-block:: chapel
+
+  proc R.init=(other : this.type) {
+    this.T = other.T;
+    this.x = other.x;
+  }
+
+The first line of this ``init=`` may seem unnecessary, since ``this.type`` must
+already be known. The line ``this.T = other.T`` currently serves as an
+assertion, and future releases may remove the need for such field
+initializations within a generic ``init=``.
+
+Field-Based Constraints
+^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``this.type`` query can also be used to constrain the given value based on
+generic fields. For example, consider the following generic record that simply
+wraps any given type:
+
+.. code-block:: chapel
+
+  record Wrapper {
+    type T;
+    var x : T;
+  }
+
+A simple ``init=`` for this type may try to infer ``T`` from the given value:
+
+.. code-block:: chapel
+
+  proc Wrapper.init=(value : ?T) {
+    this.T = T;
+    this.x = value;
+  }
+
+This only works as long as the desired instantiation of ``T`` and the type of
+the value match. What if a user tried to initialize a ``Wrapper(int(8))`` from
+an integer literal?
+
+.. code-block:: chapel
+
+  var x : Wrapper(int(8)) = 5;
+
+The type of ``5`` is actually ``int(64)``, and the ``init=`` would fail at the
+line ``this.T = T;``. Furthermore, because ``value`` is a fully-generic
+argument this ``init=`` would also resolve as the copy initializer, and attempt
+to initialize some sort of nested ``Wrapper(Wrapper(int(8)))`` type.
+
+A better approach is to constrain ``value`` using ``this.type``:
+
+.. code-block:: chapel
+
+  proc Wrapper.init=(value : this.type.T) {
+    this.T = value.type;
+    this.x = value;
+  }
+
+  var x : Wrapper(int(8)) = 5; // x.init=(5)
+  var y = x; // compiler-generated init=
+
+The literal ``5`` will now coerce from ``int(64)`` to ``int(8)`` following
+regular Chapel semantics, and the compiler-generated ``init=`` will be invoked
+when initializing variable ``y``.
+
+Using 'this.type' Inside 'init='
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A type may be initialized from a value that represents only part (or none) of
+the required instantiation information. For example, consider initializing a
+distributed list type from an array:
+
+.. code-block:: chapel
+
+  record DistList {
+    type DistType;
+    type eltType;
+    // ...
+  }
+
+  proc DistList.init=(arr : [] ?eltType) {
+    this.DistType = this.type.DistType; // from variable declaration
+    this.eltType = eltType; // from 'arr'
+    // ... initialize data, etc. ...
+  }
+
+  // Initializing a Block-distributed list from an array literal
+  var x : DistList(Block(1), int) = [1, 2, 3, 4, 5];
+
+In this example snippet, ``this.type`` is used within the ``init=`` body in
+order to achieve the desired instantiation. Part of what was needed was
+available from the given value (i.e. the element type), but the rest was
+taken from ``this.type``.
```

Comment:
It seems like it should say what would happen if you did e.g.  `var y: DistList = [1, 2, 3]`. Does that result in an error today? Can you explain how the compiler handles `this.type.DistType` being unknown? ? Or is it future work? Or?

Reply:
I tried to handle this in the next section:
```
If the value is not a subtype of the generic expression, then there will be a
compile-time error. This may change in future releases.
```

Does that work?

Reply:
Oh, I see. I think it'd be nice, in this section, to point out that only fully concrete types can be initialized in this way.